rfcs/text/remove-support-for-uncontrolled-inputs.md

```diff
@@ -0,0 +1,233 @@
+const MyComponent = ({ value, onChange, ...rest }) => {
+   const handleOnChange = (event) => {
+    if (onChange) {
+      onChange(event);
+    } 
+  };
+
+  /* Component logic goes here */
+
+  return <input value={value} onChange={handleOnChange} />;
+};
+
+MyComponent.propTypes = {
+  value: PropTypes.string.isRequired,
+  onChange: PropTypes.function.isRequired,
+};
+
+export default MyComponent;
+```
+Here is an example of how a basic component will look in TS without uncontrolled support.
+
+```tsx
+// my-component.component.tsx
+interface MyComponentProps = {
+  value: string;
+  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
+};
+
+const MyComponent = ({ value, onChange, ...rest }: myComponentProps) => {
+   const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (onChange) {
+      onChange(event.target.value);
+    } 
+  };
+
+  /* Component logic goes here */
+
+  return <input value={value} onChange={handleOnChange} />;
+};
+
+export default MyComponent;
+```
+# Detailed design
+
+- We will add deprecation warnings to our input components so that our consumers are aware of the pending changes to remove uncontrolled support. This will appear as a console warning if our consumer is using the uncontrolled mode and a warning on our Storybook documentation.
+
+- We will remove uncontrolled behaviour from our input components and associated props. This will be done in one Pull Request with each component having it's own commit message with a `BREAKING CHANGE`. The benefit of this is that we will only release one major version of Carbon. 
+
+This approach will provide immediate benefit to developers of Carbon as it means we will have less code and tests to maintain. It will also mean that consumers of Carbon can steadily upgrade their version of Carbon.
+
+## Components
+
+The input components that currently support being uncontrolled are: 
+1. Textbox
+1. Decimal
+1. Number Input
+1. Simple Select
+1. Filterable Select
+1. Multi Select
+1. Grouped Character
+1. Numeral Date
+1. Search
+1. Checkbox 
+1. RadioButton 
+1. Switch 
+1. SimpleColorPicker 
+1. AdvancedColorPicker 
+1. ButtonToggle
+
+## Unit Testing
+
+Currently uncontrolled and controlled behaviours are unit tested. As part of this work we could reduce the amount of code in the spec files quite considerably. If we take `Decimal` as an example of this, `Decimal` currently has 1430 lines of code in the spec file. This could be reduced to down to around 480 lines of code if we were to remove the uncontrolled tests. This would make the tests easier to read, maintain and reduce the runtime of the tests too.
+
+# Drawbacks
+
+- This work will be a `BREAKING CHANGE` however we will communicate this in the appropriate Carbon channels. Consumers of Carbon will also have the opportunity to comment on this RFC.
+
+- Any consumers of Carbon that currently use uncontrolled components will have to rewrite their form implementations.  
```

Comment:
I can't really explain why but `rewrite their form implementations` sounds a bit clumsy to me. I would just say `will have to rewrite their application` (and perhaps add `to use controlled components instead` to the end but that's not really necessary).

Reply:
To me `rewrite their application` implies that the whole app will need rewriting because of this change. If I were a consumer of Carbon reading that, I'd be quite alarmed.

Reply:
OK fair enough. Perhaps then, instead of "rewrite...", say something like "update their uncontrolled components to be controlled".

But the original isn't bad, I think it's just a weird unexplainable personal thing that it reads a bit weird to me. So feel free to leave this as is.

Reply:
I'll leave as is if that is OK? Sam suggest this in this comment.
https://github.com/Sage/carbon/pull/5029#discussion_r859749087

Reply:
Fair enough. I don't think my suggested change contradicts Sam's, I agree with him that we should have a note about consumers having to rewrite, I just think this can be phrased a bit better. But honestly, this was the one of my comments I felt least strongly about so happy to let it go :)